src/main/daml/Daml/Finance/Settlement/Instruction.daml

```diff
@@ -90,35 +90,43 @@ template Instruction
         assertMsg ("approval must be new" <> messageSuffix) $ approval /= this.approval
         -- clean up if previously approved
         case this.approval of
-          TakeDelivery receiverAccount -> undiscloseA step.sender receiverAccount
+          TakeDelivery receiverAccount -> undiscloseA receiverAccount
           a -> pure None
         -- disclose approval
         case approval of
           TakeDelivery receiverAccount -> do
-            discloseA step.receiver receiverAccount
+            discloseA receiverAccount
             pure ()
           _ -> pure ()
         -- update instruction state
         toInterfaceContractId <$> create this with signed = insert step.receiver signed; approval
-      execute =
+      execute Instruction.Execute{actors; undiscloseSenderAccount; undiscloseReceiverAccount} = do
+        assertMsg "actors must be subset of settlers" $ isSubsetOf actors settlers
         -- execute instruction
         case (allocation, approval) of
           (Unallocated, _) -> abort $ "instruction must be allocated" <> messageSuffix
           (_, Unapproved) -> abort $ "instruction must be approved" <> messageSuffix
-          (Pledge transferableCid, TakeDelivery account) -> do
+          (Pledge transferableCid, TakeDelivery receiverAccount) -> do
             transferable <- fetch transferableCid
-            transferableCid <- exercise transferableCid Transferable.Transfer with newOwnerAccount = account
-            Some <$> discloseT step.receiver transferableCid
+            let senderAccount = getAccount transferable
+            transferableCid <- exercise transferableCid Transferable.Transfer with newOwnerAccount = receiverAccount
+            if undiscloseSenderAccount then do undiscloseA senderAccount; pure () else pure ()
+            if undiscloseReceiverAccount then do undiscloseA receiverAccount; pure () else pure ()
```

Comment:
I think you can use `when` from the DA.Action package (see [here](https://docs.daml.com/daml/stdlib/DA-Action.html#function-da-action-when-35467))

Something akin to (you may need to add `pure ()` as the last statement) :

```suggestion
            when undiscloseSenderAccount $ undiscloseA senderAccount
            when undiscloseReceiverAccount $ undiscloseA receiverAccount
```

Reply:
changed it to `when undiscloseSenderAccount $ undiscloseA senderAccount $> ()` as discussed